### PR TITLE
Add support for Podman

### PR DESCRIPTION
**Pull request by empovit**
Add podman as an option for building images and running on Kind

### CODE REVIEW

File: deployments/container/Makefile [hunk 1]
```diff
@@ -28,7 +28,9 @@ IMAGE = $(IMAGE_NAME):$(IMAGE_TAG)
 
 ubuntu22.04: BASE_IMAGE = docker.io/ubuntu:22.04
 
-# Use a generic build target to build the relevant images
+ubi9: BASE_IMAGE = registry.access.redhat.com/ubi9/ubi-micro:latest
```
Comment (aojea) [inline deployments/container/Makefile:31]: is not good to use latest as it can have random behavior, do you have a static tag?

Comment (empovit) [reply deployments/container/Makefile]: Thanks, pinned to 9.4

File: demo/scripts/common.sh [hunk 2]
```diff
@@ -52,3 +52,5 @@ SCRIPTS_DIR="$(cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd)"
 # The name of the kind image to build / run
 : ${KIND_IMAGE:="kindest/node:${KIND_K8S_TAG}"}
 
+# Container tool, e.g. docker/podman
+: ${DOCKER:="docker"}
```
Comment (elezar) [inline demo/scripts/common.sh:56]: Not critical for this PR. Is there a more generic term that we can use for `DOCKER`? (may affect more files, but it may be good to do as a follow up)

File: demo/scripts/load-driver-image-into-kind.sh [hunk 3]
```diff
@@ -27,6 +27,16 @@ set -o pipefail
 
 source "${CURRENT_DIR}/common.sh"
 
-kind load docker-image \
-	--name "${KIND_CLUSTER_NAME}" \
-	"${DRIVER_IMAGE}"
+# Work around kind not loading image with podman
+if [[ "${DOCKER}" == "podman" ]]; then
+	IMAGE_ARCHIVE=driver_image.tar
```
Comment (elezar) [inline demo/scripts/load-driver-image-into-kind.sh:32]: should we use a file in `/tmp`.

Also in theory this workflow can be used for both runtimes. Should we just use the one instead of having a conditional block?

File: Makefile [hunk 4]
```diff
@@ -146,16 +146,21 @@ generate-crds: vendor
 			docker; \
 	fi
 
+ifeq ($(DOCKER),podman)
+DOCKER_OPTS=-v $(PWD):$(PWD):Z
```
Comment (elezar) [inline Makefile:150]: Do you happen to have a doc link to describe why this change is required?

File: deployments/container/Makefile [hunk 5]
```diff
@@ -28,7 +28,9 @@ IMAGE = $(IMAGE_NAME):$(IMAGE_TAG)
 
 ubuntu22.04: BASE_IMAGE = docker.io/ubuntu:22.04
 
-# Use a generic build target to build the relevant images
+ubi9: BASE_IMAGE = registry.access.redhat.com/ubi9/ubi-micro:9.4
```
Comment (elezar) [inline deployments/container/Makefile:31]: This is not specific to using podman? Could we move this to a different PR. I would prefer to not have to maintain this in this repo and could look at allowing a user to set `BASE_IMAGE` instead to target other distributions.

The concern is keeping the tags up to date, for example.

**Review by elezar — COMMENTED** [review]
Thanks @empovit

This looks good in general. I have some comments / questions, but I don't see any show-stoppers.

File: demo/scripts/common.sh (hunk 2 above)
Comment (empovit) [reply demo/scripts/common.sh:56]: Yes, usually it's `CONTAINER_TOOL` or something along those lines. As long as you're OK with moving away from DOCKER :) Ideally all Dockerfiles and directory names should change as well.

Comment (elezar) [reply demo/scripts/common.sh:56]: I think I would be ok with it, but may require some more discussion on our end if we start renaming things outside of the environment and make variables.

What is the generic standard for `Dockerfile`? From my perspective, this is an implied API and not just a file name.

File: demo/scripts/load-driver-image-into-kind.sh (hunk 3 above)
Comment (empovit) [reply demo/scripts/load-driver-image-into-kind.sh:32]: Normally the file will be removed right after loading into Kind, so I don't think it matters.

Correct, the flow can be used for both runtimes, but I couldn't test extensively on Docker so preferred to make as little changes as possible at the moment.

File: Makefile (hunk 4 above)
Comment (empovit) [reply Makefile:150]: I can't find it right now, but `--user` works differently with podman, and `:z` (`:Z`) is required to write to a host directory via a volume.

File: demo/scripts/common.sh (hunk 2 above)
Comment (empovit) [reply demo/scripts/common.sh:56]: Sounds good.
`Containerfile` is more inclusive :)

Comment (elezar) [reply demo/scripts/common.sh:56]: To be clear, I don't think we should rename the `Dockerfile` to `Containerfile` at this stage. See https://github.com/docker/cli/issues/3251

Comment (empovit) [reply demo/scripts/common.sh]: I've replaced `docker` command with `CONTAINER_TOOL` variable.

Comment (empovit) [reply demo/scripts/common.sh]: It make the code look better than `if [[ "${DOCKER}" == "podman" ]]` :smile:

**Comment by klueska** [comment]
@empovit what do you mean by:

> NOTE: Currently, the dra-example-driver-kubeletplugin pod crashes if Kind runs as non-root/non-sudo.

Is this true for both docker and podman or just podman? How to reproduce on docker?

File: Makefile [hunk 6]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-DOCKER   ?= docker
+CONTAINER_TOOL ?= docker
```
Comment (elezar) [inline Makefile:15]: Do we want to add a note to the README on how to get started using podman?

**Comment by empovit** [comment]
> @empovit what do you mean by:
> 
> > NOTE: Currently, the dra-example-driver-kubeletplugin pod crashes if Kind runs as non-root/non-sudo.
> 
> Is this true for both docker and podman or just podman? How to reproduce on docker?

@klueska I've described the issue in https://kubernetes.slack.com/archives/C0409NGC1TK/p1721720213996169

```console
$ kubectl logs -n dra-example-driver dra-example-driver-kubeletplugin-q5khp
Defaulted container "plugin" out of: plugin, init (init)
E0723 07:23:56.322197       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x18258e0, 0x2a0aed0})
        /build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0000061c0?})
        /build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x18258e0?, 0x2a0aed0?})
        /usr/local/go/src/runtime/panic.go:770 +0x132
github.com/fsnotify/fsnotify.(*Watcher).isClosed(...)
        /build/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:176
github.com/fsnotify/fsnotify.(*Watcher).Add(0x0, {0xc000048009?, 0xc00051b4c8?})
        /build/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:240 +0x57
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*watch).update(0xc0001112d0, 0xc0001d6900, {0x0, 0x0, 0x51b720?})
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:563 +0xd1
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*Cache).refreshIfRequired(0xc000023180, 0x8?)
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:208 +0x38
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*Cache).Refresh(0xc000023180)
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:123 +0xa6
main.NewCDIHandler(0x0?)
        /build/cmd/dra-example-kubeletplugin/cdi.go:46 +0x9e
main.NewDeviceState(0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/state.go:66 +0x76
main.NewDriver.func1()
        /build/cmd/dra-example-kubeletplugin/driver.go:53 +0x85
k8s.io/client-go/util/retry.OnError.func1()
        /build/vendor/k8s.io/client-go/util/retry/util.go:51 +0x30
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x4c9379?)
        /build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:145 +0x3e
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0xc00051b920)
        /build/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:461 +0x5a
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x0?, 0xc0002ad6f8?)
        /build/vendor/k8s.io/client-go/util/retry/util.go:50 +0xa5
k8s.io/client-go/util/retry.RetryOnConflict(...)
        /build/vendor/k8s.io/client-go/util/retry/util.go:104
main.NewDriver({0x1d09b90, 0x2aa1e40}, 0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/driver.go:42 +0x170
main.StartPlugin({0x1d09b90, 0x2aa1e40}, 0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/main.go:139 +0x18b
main.newApp.func2(0xc0002abdc0?)
        /build/cmd/dra-example-kubeletplugin/main.go:113 +0x12e
github.com/urfave/cli/v2.(*Command).Run(0xc0003514a0, 0xc0002abdc0, {0xc000110070, 0x1, 0x1})
        /build/vendor/github.com/urfave/cli/v2/command.go:274 +0x93f
github.com/urfave/cli/v2.(*App).RunContext(0xc000032000, {0x1d09b90, 0x2aa1e40}, {0xc000110070, 0x1, 0x1})
        /build/vendor/github.com/urfave/cli/v2/app.go:332 +0x566
github.com/urfave/cli/v2.(*App).Run(...)
        /build/vendor/github.com/urfave/cli/v2/app.go:309
main.main()
        /build/cmd/dra-example-kubeletplugin/main.go:60 +0x3f
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x518e77]

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0000061c0?})
        /build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xcd
panic({0x18258e0?, 0x2a0aed0?})
        /usr/local/go/src/runtime/panic.go:770 +0x132
github.com/fsnotify/fsnotify.(*Watcher).isClosed(...)
        /build/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:176
github.com/fsnotify/fsnotify.(*Watcher).Add(0x0, {0xc000048009?, 0xc00051b4c8?})
        /build/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:240 +0x57
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*watch).update(0xc0001112d0, 0xc0001d6900, {0x0, 0x0, 0x51b720?})
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:563 +0xd1
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*Cache).refreshIfRequired(0xc000023180, 0x8?)
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:208 +0x38
github.com/container-orchestrated-devices/container-device-interface/pkg/cdi.(*Cache).Refresh(0xc000023180)
        /build/vendor/github.com/container-orchestrated-devices/container-device-interface/pkg/cdi/cache.go:123 +0xa6
main.NewCDIHandler(0x0?)
        /build/cmd/dra-example-kubeletplugin/cdi.go:46 +0x9e
main.NewDeviceState(0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/state.go:66 +0x76
main.NewDriver.func1()
        /build/cmd/dra-example-kubeletplugin/driver.go:53 +0x85
k8s.io/client-go/util/retry.OnError.func1()
        /build/vendor/k8s.io/client-go/util/retry/util.go:51 +0x30
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x4c9379?)
        /build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:145 +0x3e
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0xc00051b920)
        /build/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:461 +0x5a
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x0?, 0xc0002ad6f8?)
        /build/vendor/k8s.io/client-go/util/retry/util.go:50 +0xa5
k8s.io/client-go/util/retry.RetryOnConflict(...)
        /build/vendor/k8s.io/client-go/util/retry/util.go:104
main.NewDriver({0x1d09b90, 0x2aa1e40}, 0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/driver.go:42 +0x170
main.StartPlugin({0x1d09b90, 0x2aa1e40}, 0xc00051bbd0)
        /build/cmd/dra-example-kubeletplugin/main.go:139 +0x18b
main.newApp.func2(0xc0002abdc0?)
        /build/cmd/dra-example-kubeletplugin/main.go:113 +0x12e
github.com/urfave/cli/v2.(*Command).Run(0xc0003514a0, 0xc0002abdc0, {0xc000110070, 0x1, 0x1})
        /build/vendor/github.com/urfave/cli/v2/command.go:274 +0x93f
github.com/urfave/cli/v2.(*App).RunContext(0xc000032000, {0x1d09b90, 0x2aa1e40}, {0xc000110070, 0x1, 0x1})
        /build/vendor/github.com/urfave/cli/v2/app.go:332 +0x566
github.com/urfave/cli/v2.(*App).Run(...)
        /build/vendor/github.com/urfave/cli/v2/app.go:309
main.main()
        /build/cmd/dra-example-kubeletplugin/main.go:60 +0x3f
```
I didn't test with rootless Docker as it's not common to my knowledge. IMO the problem would be best dealt with as a GitHub issue, but can't open one without adding basic support for Podman first.

Tested combinations
* Works with Docker on Ubuntu as regular user
* Works with Podman on RHEL as root user
* Works with Podman on Fedora using sudo
* Crashes with Podman on Fedora as regular user

File: Makefile (hunk 6 above)
Comment (empovit) [reply Makefile:15]: Yes, that's a good idea. I will update the PR.

**Comment by elezar** [comment]
@empovit I think that's a bug in the example driver, please create an issue for it and I will have a look at it on Monday. I think we need to update the way in which we write the CDI specifications.

**Comment by klueska** [comment]
https://github.com/kubernetes-sigs/dra-example-driver/pull/52 has been merged. Please let me know if it has resolved the issue.

**Comment by empovit** [comment]
Hmm, #52 fixes the issue on 1.30, but some other changes break the Podman support :( 
This can't run because Kind's build has Docker hard-coded:

```console
KUBE_GIT_VERSION=v1.30.0 BUILD_KIND_IMAGE=true KIND_K8S_TAG=v1.31.0-rc.1 ./demo/create-cluster.sh
```

**Comment by elezar** [comment]
Glad to hear that the repo works as-is at least. Let's dig into the build failures and decide whether we can address them here or whether we need to propose changes to kind.

**Comment by klueska** [comment]
So the issue is when building the image itself -- maybe I can / should change things to just have it pull from an image hosted in the github registry until an official kind image comes out.

**Comment by empovit** [comment]
@elezar my understanding is that it's Kind/K8s policy to use only Docker, so proposing changes to Kind won't work. 

@klueska I think that would be great. Alternatively, we can mention as a limitation that Podman works only with 1.30 until a 1.31 image is available.

**Comment by klueska** [comment]
Actually -- it looks like they already published a 1.31.0 image yesterday. Let me just open a PR to update everything to that now.

**Comment by klueska** [comment]
Done, please review:
https://github.com/kubernetes-sigs/dra-example-driver/pull/53

**Comment by klueska** [comment]
I don't see anything in the PR that tells `kind` itself to use `podman`.

I expected to see something like `KIND_EXPERIMENTAL_PROVIDER=podman` in the scripts somehwere...

File: demo/scripts/load-driver-image-into-kind.sh [hunk 7]
```diff
@@ -27,6 +27,16 @@ set -o pipefail
 
 source "${CURRENT_DIR}/common.sh"
 
-kind load docker-image \
-	--name "${KIND_CLUSTER_NAME}" \
-	"${DRIVER_IMAGE}"
+# Work around kind not loading image with podman
+if [[ "${CONTAINER_TOOL}" == "podman" ]]; then
+	IMAGE_ARCHIVE=driver_image.tar
```
Comment (klueska) [inline demo/scripts/load-driver-image-into-kind.sh:32]: Can you add this to the .gitignore

**Comment by empovit** [comment]
> I don't see anything in the PR that tells `kind` itself to use `podman`.
> 
> I expected to see something like `KIND_EXPERIMENTAL_PROVIDER=podman` in the scripts somehwere...

It's auto-detected https://kind.sigs.k8s.io/docs/user/quick-start/
> The kind can auto-detect the [docker](https://www.docker.com/), [podman](https://podman.io/), or [nerdctl](https://github.com/containerd/nerdctl) installed and choose the available one. If you want to turn off the auto-detect, use the environment variable KIND_EXPERIMENTAL_PROVIDER=docker, KIND_EXPERIMENTAL_PROVIDER=podman or KIND_EXPERIMENTAL_PROVIDER=nerdctl to select the runtime.

**Comment by klueska** [comment]
OK, so I can override it with `KIND_EXPERIMENTAL_PROVIDER` if I want, but otherwise it will be autodetected?

File: demo/scripts/load-driver-image-into-kind.sh (hunk 7 above)
Comment (empovit) [reply demo/scripts/load-driver-image-into-kind.sh:32]: Thanks, done.

**Comment by empovit** [comment]
Correct.

**Comment by klueska** [comment]
Can we instead explicitly set KIND_EXPERIMENTAL_PROVIDER to the value of CONTAINER_TOOL and instead autodetect CONTAINER_TOOL in this environment?

**Comment by empovit** [comment]
Done. PTAL

File: demo/scripts/load-driver-image-into-kind.sh [hunk 8]
```diff
@@ -27,6 +27,16 @@ set -o pipefail
...
+	${CONTAINER_TOOL} save -o "${IMAGE_ARCHIVE}" "${DRIVER_IMAGE}" && \
+	kind load image-archive \
+		--name "${KIND_CLUSTER_NAME}" \
+		"${IMAGE_ARCHIVE}"
+	rm "${IMAGE_ARCHIVE}"
+else
+	kind load docker-image \
+		--name "${KIND_CLUSTER_NAME}" \
+		"${DRIVER_IMAGE}"
+fi
```
Comment (klueska) [inline demo/scripts/load-driver-image-into-kind.sh:42]: Do these also need `KIND_EXPERIMENTAL_PROVIDER=${CONTAINER_TOOL} ` in front of them?

Comment (klueska) [reply demo/scripts/load-driver-image-into-kind.sh:42]: While we're at it, let's just drop the if/else and do it the same for both podman and docker (i.e. the new podman way).

Comment (klueska) [reply demo/scripts/load-driver-image-into-kind.sh:42]: Also, mabey the better way to do this is to introduce a `KIND` variable instead of calling `kind` directly. Then in common.sh we define `KIND="KIND_EXPERIMENTAL_PROVIDER=${CONTAINER_TOOL} kind"` and replace all direct calls to `kind` with `${KIND}`.

Comment (empovit) [reply demo/scripts/load-driver-image-into-kind.sh:42]: > While we're at it, let's just drop the if/else and do it the same for both podman and docker (i.e. the new podman way).

I can't currently test it. See my earlier  response.

Comment (empovit) [reply demo/scripts/load-driver-image-into-kind.sh:42]: > Do these also need `KIND_EXPERIMENTAL_PROVIDER=${CONTAINER_TOOL} ` in front of them?

According to the docs, it only works for `kind create cluster` (looking at the code seems to confirm that), will be ignored with all other commands. We still can implement your suggestion about `${KIND}` variable, hopefully this won't cause confusion.

Comment (empovit) [reply demo/scripts/load-driver-image-into-kind.sh:42]: Ok, I've made the requested changes, also tested on Ubuntu+Docker. PTAL

File: demo/scripts/common.sh [hunk 9]
```diff
@@ -55,3 +55,18 @@ SCRIPTS_DIR="$(cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd)"
...
+    elif [[ -n "$(which podman)" ]]; then
+        echo "Podman found in PATH."
+        CONTAINER_TOOL=podman
+    else
+        echo "No container tool detected. Please install Docker or Podman."
+        return 1
+    fi
+fi
+
+: ${KIND:="env KIND_EXPERIMENTAL_PROVIDER=${CONTAINER_TOOL} kind"}
```
Comment (klueska) [inline demo/scripts/common.sh:72]: just for my own understanding -- why is the `env` included here?

Comment (empovit) [reply demo/scripts/common.sh:72]: Right, without it substituting `${KIND}` makes the shell treat `KIND_EXPERIMENTAL_PROVIDER=podman` as a command, causing something like `.../demo/scripts/create-kind-cluster.sh: line 30: KIND_EXPERIMENTAL_PROVIDER=podman: command not found`.

If you can think of a better solution, I'll be happy to use it.

Comment (klueska) [reply demo/scripts/common.sh:72]: I guess the alternative would be to make sure that `KIND_EXPERIMENTAL_PROVIDER` is exported in every script where `kind` is called (and then go back to not having a `KIND` variable at all).

Comment (klueska) [reply demo/scripts/common.sh:72]: To be clear, I don't feel too strongly either way so long as it works (and will work in all supported environments).

Comment (empovit) [reply demo/scripts/common.sh:72]: I personally see no harm in using `env` as its purpose is exactly running a command with passed env variables https://www.gnu.org/software/coreutils/manual/html_node/env-invocation.html. It's also widely supported https://en.wikipedia.org/wiki/Env. On the other hand, it allows us to set the `KIND` variable in a single place, and keep it alongside other Kind-related variables.

**Comment by klueska** [comment]
@empovit thanks for all the iterations on this. I've tested it on my side with both podman and docker and everything seems to work as expected.

/lgtm
/approve
